synapse/types.py

```diff
@@ -63,6 +64,10 @@
 # JSON types. These could be made stronger, but will do for now.
 # A JSON-serialisable dict.
 JsonDict = Dict[str, Any]
+# A JSON-serialisable mapping; roughly speaking an immutable JSONDict.
+# Useful when you have a TypedDict which isn't going to be mutated and you don't want
+# to cast to JsonDict everywhere.
+JsonMapping = Mapping[str, Any]
```

Comment:
It's worth noting that `Mapping`s aren't necessarily JSON-serialisable, unless they're really `dict`s or `frozendict`s. Essentially when we use `JsonMapping`, we're pinky promising that it's really a `(frozen)dict`.

The same kind of thing applies to `JsonSerializable` below I guess.

Reply:
I wasn't 100% sold on this either. I wanted to see if there was a way to make more use of `TypedDicts` without having to cast those back to `JsonDicts` everywhere. Maybe I could just make the return type of this function `SearchResult`? That feels a bit off though---it's exposing a type from storage.databases.main in rest.client.

https://github.com/matrix-org/synapse/blob/7fe7b42e4b3f7c9b0172bc79dd84deafc94a23b7/synapse/rest/client/user_directory.py#L41

Reply:
I do think there is a case for having a `JsonMapping` type, to signal that functions aren't allowed to modify the contents of a dict. Especially around caches.

I'm not _opposed_ to making the return type some sort of `TypedDict`. I think we really ought to have a file somewhere full of `TypedDict` definitions that come from the spec, since that's the source of truth for the return type here (and not the database layer).

In any case I'm happy with the `JsonMapping` as is.

Reply:
> I do think there is a case for having a `JsonMapping` type, to signal that functions aren't allowed to modify the contents of a dict. Especially around caches.

I broadly sympathise, but I don't think that'd be fool proof: e.g. given `d: JsonMapping = {"a": [1,2,3]}` I think `d["a"].append(4)` would still type check?

> I think we really ought to have a file somewhere full of TypedDict definitions that come from the spec, since that's the source of truth for the return type here (and not the database layer).

I like the sound of this; e.g. `synapse.types.spec.UserDirectorySearchResult`.  I guess the dream would be to automatically generate such a file from the openapi definition in the spec.

Reply:
> > I do think there is a case for having a `JsonMapping` type, to signal that functions aren't allowed to modify the contents of a dict. Especially around caches.
> 
> I broadly sympathise, but I don't think that'd be fool proof: e.g. given `d: JsonMapping = {"a": [1,2,3]}` I think `d["a"].append(4)` would still type check?

True! Unless we manage to define `JsonMapping = Mapping[str, ImmutableJsonSerializable]` recursively, which mypy won't let us do. (grumble grumble this would be possible in typescript)